facilitator/src/aggregation.rs

```diff
@@ -109,10 +110,10 @@ impl<'a> BatchAggregator<'a> {
     /// Compute the sum part for all the provided batch IDs and write it out to
     /// the aggregation transport. The provided callback is invoked after each
     /// batch is aggregated.
-    pub fn generate_sum_part<F: FnMut() + Send + Clone + 'static>(
+    pub fn generate_sum_part<W: Watchdog>(
         &mut self,
         batches: Vec<SumInput>,
-        callback: F,
+        watchdog: &'a W,
```

Comment:
The difference is that `generate_sum_part` is multithreaded, while `BatchIntaker::generate_validation_share` is not. The latter calls `send_heartbeat()` directly on the `watchdog` it is passed while `generate_sum_part` uses `clone()` to safely send "copies" of the watchdog to each of the threads it spawns. the `'a` lifetime is a leftover from an earlier iteration where I was struggling to get the lifetimes right, and I've deleted it.

I agree that it's odd that the two method signatures don't match, and the reason why leaks the detail that one method is multithreaded and the other isn't. What do you think can be done to improve this? Take `&W: Watchdog` in both methods and always clone into a mutable instance, just for consistency?

Reply:
Yeah, I would treat `Watchdog` as a shared reference in all cases. As a side note, you don't necessarily have to clone the Watchdog to get a mutable instance; it might make more sense for your Watchdogs that actually mutate state (i.e. the ones that extend a Task's deadline) to use interior mutability for it.

Reply:
The `TaskQueueWatchdog` already wraps its interesting fields in `Arc+Mutex` so I think that already achieves interior mutability. Put another way, I'm not sure what more would be achieved by sticking a `RefCell` somewhere between `Arc` and `Mutex`.

Sure enough, I found that I was able to just remove the `mut` from the `send_heartbeat(&mut self);` method on the `Watchdog` trait and everything Just Works, since `std::sync::Mutex::lock` takes `&self`.